File: implementations/typescript/packages/event/src/router.js
import { matchesType } from "./subscription.js";

export class EventRouter {
  constructor() {
    this._handlers = [];
  }

  on(patternsOrHandler, handler) {
    if (handler === undefined) {
      this._handlers.push({ match: () => true, handler: patternsOrHandler });
    } else if (typeof patternsOrHandler === "function") {
      this._handlers.push({ match: patternsOrHandler, handler });
    } else {
      const patterns = Array.isArray(patternsOrHandler) ? patternsOrHandler : [patternsOrHandler];
      this._handlers.push({ match: (event) => patterns.some((pattern) => matchesType(pattern, event.type)), handler });
    }
    return this;
  }

  dispatch(event) {
    const results = [];
    for (const { match, handler } of this._handlers) {
      if (!match(event)) continue;
      const responseOrResponses = handler(event);
      if (responseOrResponses !== undefined && responseOrResponses !== null) {
        results.push(...(Array.isArray(responseOrResponses) ? responseOrResponses : [responseOrResponses]));
      }
    }
    return results;
  }

  static pattern(pattern) {
    return { forSession: (sessionId) => (event) => matchesType(pattern, event.type) && event.session_id === sessionId };
  }
}

File: implementations/python/src/axisrobo_harmovela_event/router.py
from .subscription import matches_type


class EventRouter:
    def __init__(self):
        self._handlers: list[tuple] = []

    def on(self, patterns_or_fn, handler=None):
        if handler is None:
            self._handlers.append((lambda _: True, patterns_or_fn))
        elif callable(patterns_or_fn):
            self._handlers.append((patterns_or_fn, handler))
        else:
            patterns = [patterns_or_fn] if isinstance(patterns_or_fn, str) else patterns_or_fn
            self._handlers.append(
                (lambda event, values=patterns: any(matches_type(pattern, event.get("type")) for pattern in values), handler)
            )
        return self

    def dispatch(self, event: dict) -> list:
        results = []
        for match_fn, handler in self._handlers:
            if match_fn(event):
                response = handler(event)
                if response is not None:
                    results.extend(response if isinstance(response, list) else [response])
        return results
